### PR TITLE
Resolve the Flaky Test - TestWaitCustomTask_PipelineRun

### LINKED ISSUE #5653: PipelineRun reconciler shouldn't care about how the `Timeout` field is implemented in Custom Task Run.
PipelineRun reconciler shouldn't care about how the Timeout field is implemented in Custom Task Run.

Currently in PipelineRun reconciler, it cancells the Run when it detects that Run is TimeOut: https://github.com/tektoncd/pipeline/blob/2b54123eaafe5d6b86577402830e0957928374d2/pkg/reconciler/pipelinerun/pipelinerun.go#L657-L659
by calling https://github.com/tektoncd/pipeline/blob/2b54123eaafe5d6b86577402830e0957928374d2/pkg/apis/pipeline/v1alpha1/run_types.go#L230-L241

This behavior leads to flaky unit test results now and then. See the test result of `TestWaitCustomTask_PipelineRun/Wait_Task_Retries_on_Timeout` in 
- https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5547/pull-tekton-pipeline-alpha-integration-tests/1574802170165858304/build-log.txt
- https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/5648/pull-tekton-pipeline-alpha-integration-tests/1582078450888347648/build-log.txt

Say we have defined the following PipelineRun
```yaml
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: custom-task-pipeline
spec:
  tasks:
  - name: run-wait
    timeout: "1s"
    retries: 1
    taskRef:
      apiVersion: wait.testing.tekton.dev/v1alpha1
      kind: Wait
    params:
    - name: duration
      value: "2s"
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: custom-task-pipelinerun
spec:
  pipelineRef:
    apiVersion: tekton.dev/v1beta1
    name: custom-task-pipeline
```

# Expected Behavior
1. PR reconciler creates a Run: `custom-task-pipelinerun-run-wait` https://github.com/tektoncd/pipeline/blob/8399472eda2706e5416751b74daae7af22ab75ab/pkg/reconciler/pipelinerun/pipelinerun.go#L790-L801
2. Run reconciler reconcils the Run, continues reconciling because the Run hasn't TimedOut.
3. Run reconciler detects the Run has timed out, marks it as failed on `TimedOut` after `1s`, example: https://github.com/tektoncd/pipeline/blob/8399472eda2706e5416751b74daae7af22ab75ab/test/custom-task-ctrls/wait-task/pkg/reconciler/reconciler.go#L126-L139
4. PR reconciler detects the Run is marked as `TimedOut`, update its `ChildReferences` or `RunsStatus`: https://github.com/tektoncd/pipeline/blob/8399472eda2706e5416751b74daae7af22ab75ab/pkg/reconciler/pipelinerun/pipelinerun.go#L684-L691

# Actual Behavior
1. PR reconciler creates a Run: `custom-task-pipelinerun-run-wait` https://github.com/tektoncd/pipeline/blob/8399472eda2706e5416751b74daae7af22ab75ab/pkg/reconciler/pipelinerun/pipelinerun.go#L790-L801
2. Run reconciler reconcils the Run, continues reconciling because the Run hasn't TimedOut.
3. **[The flaky part]** PR reconciler `processRunTimeouts` https://github.com/tektoncd/pipeline/blob/2b54123eaafe5d6b86577402830e0957928374d2/pkg/reconciler/pipelinerun/pipelinerun.go#L657-L659 at this time, **The Run has TimedOut, but the Run reconciler hasn't update the Run's status yet**, then PR reconciler tries to cancel the Run https://github.com/tektoncd/pipeline/blob/2b54123eaafe5d6b86577402830e0957928374d2/pkg/reconciler/pipelinerun/pipelinerun.go#L718-L725

### PR DESCRIPTION
**Pull request by XinruZhang**
# Changes
Fix #5653 

I'm openning this PR as a tentative solution to address #5653:

**TL;DR**: `TestWaitCustomTask_PipelineRun/Wait_Task_Retries_on_Timeout` has been flaky for a while. This PR stops the PipelineRun reconciler from cancelling Run when it detects that the **task-level Timeout** configured for the Run has passed, which will address the flake (similar to https://github.com/tektoncd/pipeline/pull/5134 which addresses `TestPipelineRunTimeout`).

Some other useful information, many thanks to @abayer, @lbernick and @jerop:
- `Run` should only be canceled when
  - The `PipelineRun` as a whole has timed out, either due to `PipelineRun.Spec.Timeout`'s value or `PipelineRun.Spec.Timeouts.Pipeline`'s value.
  - It's a non-finally `PipelineTask`, `PipelineRun.Spec.Timeouts.Tasks` is set, and that value has elapsed.
  - It's a finally `PipelineTask`, `PipelineRun.Spec.Timeouts.Finally` is set, and that value has elapsed.
- For `TaskRun`, pipelinerun reconciler don't cancel it when it times out.

This change only applies to **task-level Timeout**, i.e. when `PipelineTask.Timeout` is set for a Run.

cc @pritidesai  Will this commit have any other side effect?

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
ACTION REQUIRED: Starting from this release, Custom Task Runs controllers need to implement the `Timeout` on your own,  PipelineRun reconciler would not set `Run.Spec.Status == RunCancelled` upon Run timeout.
```

### CODE REVIEW

**Comment by abayer** [comment]
> * `Run` should only be canceled when its uplevel PipelineRun is timed out

A clarification: a `Run` is cancelled for timing out when any of the following are true:
* The `PipelineRun` as a whole has timed out, either due to `PipelineRun.Spec.Timeout`'s value or `PipelineRun.Spec.Timeouts.Pipeline`'s value.
* It's a non-`finally` `PipelineTask`, `PipelineRun.Spec.Timeouts.Tasks` is set, and that value has elapsed.
* It's a `finally` `PipelineTask`, `PipelineRun.Spec.Timeouts.Finally` is set, and that value has elapsed.

**Review by jerop — COMMENTED** [review]
Thank you @XinruZhang for looking into this flake!

Please update the release notes to reflect that, with this change, implementors of Runs would have to fully implement Run timeouts themselves - PipelineRun reconciler would not set `Run.Spec.Status == RunCancelled` upon Run timeout

Might be worth also clarifying that this change applies only to task-level timeouts, which should be handled by run reconcilers for custom tasks, in the same way that taskrun reconcilers handles task-level timeouts for tasks (since https://github.com/tektoncd/pipeline/pull/5134)

Pipeline-level timeouts continue to work as is - through cancelling TaskRuns and Runs:
https://github.com/tektoncd/pipeline/blob/58fd07c2316e14ab9748c83ef06aa85f27ccfc0c/pkg/reconciler/pipelinerun/timeout.go#L129-L136

A similar change (https://github.com/tektoncd/pipeline/pull/5134) addressed flakiness in `TestPipelineRunTimeout`

cc @tektoncd/core-maintainers

**Comment by XinruZhang** [comment]
Appreciate all the info @jerop! Updated the PR description and release note.

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 1]
```diff
@@ -1634,17 +1634,9 @@ status:
...
-		Value:     "RunCancelled",
-	}, {
-		Operation: "add",
-		Path:      "/spec/statusMessage",
-		Value:     string(v1alpha1.RunCancelledByPipelineMsg),
-	}}
+	var want []jsonpatch.JsonPatchOperation
 	if d := cmp.Diff(got, want); d != "" {
-		t.Fatalf("Expected cancel patch operation, but got a mismatch %s", diff.PrintWantGot(d))
+		t.Fatalf("Expected empty operation, but got a mismatch %s", diff.PrintWantGot(d))
```
[the remark this answers is not in the document]
Comment (XinruZhang) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go]: Good point, updated, thanks!!

**Review by jerop — COMMENTED** [review]
Thank you for investigating and addressing this flake @XinruZhang! 

I support this change, but will wait for @ScrapCodes (who may depend on this functionality) to chime in - @tektoncd/core-maintainers please take a look

For future reference, the wait custom task that we use for testing has fully implemented timeouts itself (can serve as an example of not relying on pipelinerun reconciler) - https://github.com/tektoncd/pipeline/blob/b648a5b2eaebf0ece7199dcf2bd89c65137973d5/test/custom-task-ctrls/wait-task/pkg/reconciler/reconciler.go#L101-L139

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 2]
```diff
@@ -1772,7 +1765,7 @@ status:
 			}, {
 				Operation: "add",
 				Path:      "/spec/statusMessage",
-				Value:     string(v1alpha1.RunCancelledByPipelineMsg),
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:1775]: can this constant be removed?

Comment (XinruZhang) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:1775]: I feel It's better if we keep it same as other test cases. Oh actually, we may want to update the previous one as `v1alpha1.RunReasonCancelled`, nice catch!

Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:1775]: sorry, I meant can `RunCancelledByPipelineMsg` be removed from the v1alpha1 package?

Comment (XinruZhang) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:1775]: Oh yes, you are totally right, sorry I should've asked. We'll need to move all related usage of `v1alpha1.Run` to `v1beta1` later on as part of https://github.com/tektoncd/pipeline/issues/5153, added https://github.com/tektoncd/pipeline/issues/5153#issuecomment-1290866807 to make sure this will happen.

**Review by jerop — APPROVED** [review]
/lgtm

**Comment by jerop** [comment]
/test tekton-pipeline-unit-tests

cloud events flake

**Comment by ScrapCodes** [comment]
@jerop and @XinruZhang I was wondering, without this feature how will custom task controller "reconcile" and take action on timed-out task. Basically, this would cause that trigger.

**Comment by XinruZhang** [comment]
> 

Hi Prashant, thanks for keeping eye on this issue!

If the Run controller handles the timeout correctly -- mark the Run as Failed on timeout, then we wouldn't have any problem. 
If the Run controller doesn't implement the TimeOut functionality, the entire PipelineRun will eventually fail on pipeline-level timeout.

**Comment by ScrapCodes** [comment]
Hi Xinru,
Thanks!, that part is understood. Suppose a custom task is performing an action and is stuck for some reason. The custom task controller will not get to "reconcile" it until it is updated in someway.  
The reconcile is triggered not periodically but whenever there is an update. I need to test this out, to be sure.

**Comment by XinruZhang** [comment]
Oh I see, great point! After thinking more about it, I think that would be a concern of Custom Run controller right? A custom controller can requeue a key after a period of time, similar to
https://github.com/tektoncd/pipeline/blob/db4c05c0294ad33a2f6e7b17cacf53526139616d/test/custom-task-ctrls/wait-task/pkg/reconciler/reconciler.go#L108-L116

**Comment by ScrapCodes** [comment]
Is this scalable? when number of objects are large. Do you think the overhead of examining a large number of run object by custom task controller, every object's "timeout time" can impact performance?
